liquibase-standard/src/main/java/liquibase/changelog/visitor/UpdateVisitor.java

```diff
@@ -99,8 +94,13 @@ private static void upgradeCheckSumVersionForAlreadyExecutedOrNullChange(ChangeS
             Scope.getCurrentScope().getUI().sendMessage(String.format("Upgrading checksum for Changeset %s from %s to %s.",
                     changeSet, oldChecksum, changeSet.getStoredCheckSum()));
         }
-        Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("jdbc", database)
-                .execute(new UpdateChangeSetChecksumStatement(changeSet));
+        ChangeLogHistoryService changeLogService = Scope.getCurrentScope().getSingleton(ChangeLogHistoryServiceFactory.class).getChangeLogService(database);
+        if (changeLogService instanceof AbstractChangeLogHistoryService) {
```

Comment:
Doing `if instanceof` is usually a bad code smell. The big reason for OOP is to be making the calling code not care about the implementation details and only tell the interface what it wants to do.

In this case, nothing outside of the HistoryService implementations should know or care about UpdateChangeSetChecksumStatement, that should just be an implementation detail of some subclasses of the interface. From the standpoint of this UpdateVisitor, it should just be calling a descriptive method of what it wants from whatever the HistoryService is, like `changeLogService.replaceChecksum(changeSet)` and it's up to the implementation to do what you want.

That is a new method for the interface so you have to decide if you want that to be a breaking API change, or if you want to put a default implementation on the interface. Maybe one that executes the UpdateChangeSetChecksumStatement?

The job of the HistoryService is to wrap ALL the logic for managing the history because that is how extensions control it. You can't even assume a HistoryService is even built around statements, which is why the logic needs to be around a method not on a Generator.

But, you should really try to avoid breaking API changes, and default interface impementations are your friends for that. Like with 4.23 you maybe could have added defaults for the new getRanChangeSets(boolean) and isDatabaseChecksumsCompatible() that call off to the old methods and/or returned "fallback" values like "false"?

Reply:
Thanks for the review @nvoxland ! 
I tried moving it to ChangeLogHistoryService but then I need another method to get the database, and this one would need to be abstract breaking the interface:

```
    Database getDatabase();

    default void replaceChecksum(ChangeSet changeSet) throws DatabaseException {
        Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("jdbc", getDatabase()).execute(new UpdateChangeSetChecksumStatement
                (changeSet));
        getDatabase().commit();
        reset();
    }
```

We can also leave it as is now just using UpdateChangeSetChecksumStatement and don't touch ChangeLogHistoryService, but seems it is hard for extensions to implement it. I don't know which is worse.

Reply:
True, you would need that getDatabase() do do the default. I don't think you can/should rely on the `UpdateChangeSetChecksumStatement` for updating the checksums, it needs to be a method on the HistoryService. 

And you don't want to add `getDatabase()` to the HistoryService interface because I don't want to be forcing/requiring a single Database on the HistoryService like that. You want to keep the interface as clean and generic as possible.

So it does seem like it is a new method that's needed which unfortunately will then be a breaking API change, unless there's a default implementation that you can come up with that doesn't require any other methods to be added. Having the default be a no-op wouldn't be a valid default, would it?

Reply:
The default no-op is a good idea! I added it to the other method too. What do you think now?